DataBiosphere/jade-data-repo
# [DR-2625] Check for RAS passport in snapshot enumerate endpoint

https://broadworkbench.atlassian.net/browse/DR-2625
Spike doc: https://docs.google.com/document/d/1dD3HgPOIKxdH41fkN4QbLFFipytPBshCnmsEj8mGmro/edit#heading=h.1yvpzjf3b7q0

**Objective**

 A linked RAS passport indirectly establishes its holder as a reader of snapshots matching its decoded passport criteria -- PHS ID and consent code.

Calls to [enumerateSnapshots](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/snapshots/enumerateSnapshots) endpoint now include any snapshots available to the caller via their linked RAS passport if one exists.

Previously, only permissions directly established by SAM were considered when enumerating available passports.

**Changes**

I broke changes up into meaningful commits.  In response to PR review, a later commit involves a fair amount of renaming, so I think review will be easier to step through commit by commit.

- `ECMService.getRASDbgapPermissions ` fetches the caller's linked passport, decodes it to get its embedded tokens (visas), then decodes those visas to get their `RASDbgapPermissions` (PHS ID and consent code)
- `SnapshotDAO.getAccessibleSnapshots(permissions)` returns snapshot UUIDs accessible under those passed-in `RASDbgapPermissions`
- `SnapshotService.listAuthorizedSnapshots` combines existing SAM-accessible snapshots with new RAS-acceptable snapshots.  This is passed to `SnapshotService.enumerateSnapshots`, and all code downstream including DAO interactions remain unaltered.

NOTE: the team discussed whether audit trail logging was required.   As we are not reading underlying snapshot data and only accessing its metadata, we decided that it wasn't. However, other tickets stemming from the same Spike doc may require audit trail logging (ex. using RAS passports to fetch snapshots). Requirements: https://docs.google.com/document/d/1isNIYTkoOM0TdQ3skcSu79VMh8CCPHq_Zh356VXSVDE/edit#heading=h.q6aelx7la7n1

**Manual Verification**

My [dev environment](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html) is up to date with these changes.

Snapshot [6510b127-c226-46a1-ac12-c5f2383cbd08](https://jade-ok.datarepo-dev.broadinstitute.org/snapshots/6510b127-c226-46a1-ac12-c5f2383cbd08) has `phsId=phs000710` and `consentCode=c99`.  It is directly accessible only by my test user as a steward (and under the hood by admins):

<img width="1053" alt="Screen Shot 2022-06-28 at 9 10 26 PM" src="https://user-images.githubusercontent.com/79769153/176329860-77fd3151-dc01-4983-9aa8-c64caa074a19.png">

As my non-admin user okotsopo@broad, I logged into Terra Dev, then navigated to [Profile](https://bvdp-saturn-dev.appspot.com/#profile) > RAS > Link your RAS account.  I then logged into the NIH portal as user [Broadtestuser115](https://docs.google.com/spreadsheets/d/13jaYEFeQgUV0j2QqfwOCjw1h9APrrAxs5xbH60BW6oo/edit#gid=0) and observed that my passport had been successfully linked.  That passport includes permissions matching those in the snapshot above.

When enumerating snapshots as okotsopo@broad, I now see the passport-accessible snapshot returned with `reader` role as expected, even though my user does not have direct `reader` privileges from SAM:

```
{
  "total": 1,
  "filteredTotal": 1,
  "items": [
    {
      "id": "6510b127-c226-46a1-ac12-c5f2383cbd08",
      "name": "DR2472_person_sample_PHS_consentcode_snap",
      "description": "Snapshot with PHS and consent code",
      "createdDate": "2022-05-09T17:49:28.279008Z",
      "profileId": "8c3aeb0c-8de1-4cdd-92ed-8e03c05acbc9",
      "storage": [
        {
          "region": "us-central1",
          "cloudResource": "bigquery",
          "cloudPlatform": "gcp"
        },
        {
          "region": "us-east4",
          "cloudResource": "firestore",
          "cloudPlatform": "gcp"
        },
        {
          "region": "us-central1",
          "cloudResource": "bucket",
          "cloudPlatform": "gcp"
        }
      ],
      "secureMonitoringEnabled": false,
      "consentCode": "c99",
      "phsId": "phs000710",
      "cloudPlatform": "gcp",
      "dataProject": "datarepo-dev-dc861ff7",
      "storageAccount": null,
      "selfHosted": false
    }
  ],
  "roleMap": {
    "6510b127-c226-46a1-ac12-c5f2383cbd08": [
      "reader"
    ]
  }
}
```

If you are interested, you can try this too by linking the same passport in Terra Dev and enumerating snapshots in jade-ok environment.
